File: 0576 - Out of Boundary Paths/solution.c
int findPaths(int m, int n, int maxMove, int startX, int startY){
    int res = 0, MOD = 1e9+7, dir[5] = {1, 0, -1, 0, 1}, dp[2][52][52] = {}, N = 1;
    dp[0][startX+1][startY+1] = 1;
    for(int i = 1; i <= maxMove; i++){
        for(int x = 1; x <= m; x++){
            for(int y = 1; y <= n; y++){
                dp[N][x][y] = 0;
                for(int j = 0; j < 4; j++)
                    dp[N][x][y] = (dp[N][x][y]+dp[N^1][x+dir[j]][y+dir[j+1]])%MOD;
                long long cnt = (x==1)+(x==m)+(y==1)+(y==n);
                res = (res+cnt*dp[N^1][x][y])%MOD;
            }
        }
        N ^= 1;
    }
    return res;
}

File: 0576 - Out of Boundary Paths/solution.py
class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startX: int, startY: int) -> int:
        @lru_cache(None)
        def dp(step, x, y):
            if not (0 <= x < m and 0 <= y < n):
                return 1
            return sum(dp(step-1, *p) for p in ((x+1,y),(x-1,y),(x,y+1),(x,y-1)))%MOD if step else 0
        MOD = int(1e9+7)
        return dp(maxMove, startX, startY)